okio/src/nativeMain/kotlin/okio/Inflater.kt

```diff
@@ -35,70 +36,86 @@ import platform.zlib.z_stream_s
+  internal val dataProcessor: DataProcessor = object  : DataProcessor() {
+    @Throws(ProtocolException::class)
+    override fun process(): Boolean {
+      check(!closed) { "closed" }
+      require(0 <= sourcePos && sourcePos <= sourceLimit && sourceLimit <= source.size)
+      require(0 <= targetPos && targetPos <= targetLimit && targetLimit <= target.size)
 
-        val targetByteCount = targetLimit - targetPos
-        zStream.next_out = when {
-          targetByteCount > 0 -> pinnedTarget.addressOf(targetPos) as CPointer<UByteVar>
-          else -> null
-        }
-        zStream.avail_out = targetByteCount.toUInt()
+      source.usePinned { pinnedSource ->
+        target.usePinned { pinnedTarget ->
+          val sourceByteCount = sourceLimit - sourcePos
+          zStream.next_in = when {
+            sourceByteCount > 0 -> pinnedSource.addressOf(sourcePos) as CPointer<UByteVar>
+            else -> null
+          }
+          zStream.avail_in = sourceByteCount.toUInt()
 
-        val inflateResult = platform.zlib.inflate(zStream.ptr, Z_NO_FLUSH)
+          val targetByteCount = targetLimit - targetPos
+          zStream.next_out = when {
+            targetByteCount > 0 -> pinnedTarget.addressOf(targetPos) as CPointer<UByteVar>
+            else -> null
+          }
+          zStream.avail_out = targetByteCount.toUInt()
 
-        sourcePos += sourceByteCount - zStream.avail_in.toInt()
-        targetPos += targetByteCount - zStream.avail_out.toInt()
+          val inflateResult = platform.zlib.inflate(zStream.ptr, Z_NO_FLUSH)
 
-        when (inflateResult) {
-          Z_OK, Z_BUF_ERROR -> {
-            return targetPos < targetLimit
-          }
+          sourcePos += sourceByteCount - zStream.avail_in.toInt()
+          targetPos += targetByteCount - zStream.avail_out.toInt()
 
-          Z_STREAM_END -> {
-            finished = true
-            return true
-          }
+          when (inflateResult) {
+            Z_OK, Z_BUF_ERROR -> {
+              return targetPos < targetLimit
+            }
+
+            Z_STREAM_END -> {
+              finished = true
+              return true
+            }
 
-          Z_DATA_ERROR -> throw ProtocolException("Z_DATA_ERROR")
+            Z_DATA_ERROR -> throw ProtocolException("Z_DATA_ERROR")
 
-          // One of Z_NEED_DICT, Z_STREAM_ERROR, Z_MEM_ERROR.
-          else -> throw ProtocolException("unexpected inflate result: $inflateResult")
+            // One of Z_NEED_DICT, Z_STREAM_ERROR, Z_MEM_ERROR.
+            else -> throw ProtocolException("unexpected inflate result: $inflateResult")
+          }
         }
       }
     }
+
+    override fun close() {
+      if (closed) return
+      closed = true
+
+      inflateEnd(zStream.ptr)
+      nativeHeap.free(zStream)
+    }
   }
 
-  override fun close() {
-    if (closed) return
-    closed = true
+  actual constructor() : this(false)
+
+  @OptIn(UnsafeNumber::class)
```

Comment:
These should be safe in practice! We don’t expose target-specific numbers in our public API